### PR TITLE
Add report to generate AFF folder

### PR DESCRIPTION
**Pull request by albertmink**
starting point of #92. Most probably there will be several PR to get the feature working. Anyhow, contributions are welcome as always :purple_heart:

### CODE REVIEW

**Comment by larshp** [comment]
double space, false positive, https://github.com/abaplint/abaplint/issues/2523 opened

bug if you fix the bad pragma placement first, then it should also fix the double space false positive

**Comment by larshp** [comment]
few parameters fixed in https://github.com/open-abap/open-abap/pull/315/commits/4f151ae30f0b6a17ae1b3c27ebf7e231dfa69ffc

**Comment by larshp** [comment]
indentation, enable `selectionScreenBlockIndentation` in https://rules.abaplint.org/indentation/

**Comment by albertmink** [comment]
Folder name is UPPER, should be LOWER ❗ 
image

**Comment by albertmink** [comment]
@larshp 
* can we tell abapLint not to lint this program, or say ignore all findings for objects `cl_aff_log`, `cl_aff_factory ` and others?
The report relies on AFF framework which is not 7.42 and is only for SAP internal usage.
* I like abapGit standalone feature a lot. Is it possible to update the package.json file to have this feature as well?

**Comment by larshp** [comment]
ignoring, add the filename in either "exclude" or "noIssues", in this case I think we can exclude it alltogether,

image

**Comment by larshp** [comment]
standalone, abapGit uses https://github.com/larshp/abapmerge / https://www.npmjs.com/package/abapmerge to build the standalone report take a look... I dont remember how it works 😄

**Comment by albertmink** [comment]
@huber-nicolas I tested the PROG with the object type FUGR and it looked good. So it might be sufficient to test it for CLAS and it should be alright.

File: src/z_generate_file_format_folder.prog.abap [hunk 1]
```diff
@@ -0,0 +1,581 @@
...
+
+
+
+
+    DATA(readme) = VALUE string_table(
+( |# { aff_object-object_type } File Format| )
+( `` )
+( `File | Cardinality | Definition | Schema | Example`)
+( `:--- | :---  | :--- | :--- | :---` )
+( |`<name>.{ aff_object-object_type }.json` \| 1 \| [`{ aff_object-interface }.intf.abap`](./type/{ aff_object-interface }.intf.abap) \| | &&
```
Comment (huber-nicolas) [inline src/z_generate_file_format_folder.prog.abap:222]: All parts need to be in lower case here

**Review by huber-nicolas — COMMENTED** [review]
I tested it with object type class and found a small thing. See my comment.

Should this report also be able to generate the schema only?

File: src/z_generate_file_format_folder.prog.abap (hunk 1 above)
Comment (albertmink) [reply src/z_generate_file_format_folder.prog.abap]: Thanks

**Comment by albertmink** [comment]
This report is decoupled from PROG z_generate_json_schema.
The latter is the basis for tooling development and automated JSON Schema generation.
The report in the pull request is a tool to upload and maintain the abap-file-formats repository and might be obsolete in the future. But for now, the development of new file formats requires some minimal tooling, that facilitates the **conversion** from some ABAP interface type (**ABAP system**) to the AFF folder (**GitHub**)

**Comment by huber-nicolas** [comment]
Do we need a small unit test class in the report?

**Comment by wurzka** [comment]
I think Nicolas is right. This report also shall be possible to generate only a schema and only a simple transformation as object type owner will need this for now.

**Comment by albertmink** [comment]
@wurzka for XSLT and JSON Schema there is still this PROG (suitable for maintaining the AFF in the ABAP system)
image
Is this ok for the object type owner to have two reports?

**Comment by wurzka** [comment]
No, I don't think so. This report is not very powerful. It doesn't have any error log. I think we decided to bring our internal report to this repository

**Comment by albertmink** [comment]
I will close this pull request and move to another pull request that will contain the full blown report as-is from our ABAP system. The features there are necessary to accomplish the developers needs and comfort.

Changes here, will still be available since the branch won't be deleted immediately.

**Pull request closed without merging**